website/home/_posts/wheres_the_graph.md

```diff
@@ -0,0 +1,148 @@
+
+## The Graph Trap
+
+I used to be that developer drawing elaborate workflow diagrams on whiteboards. Building complex orchestration pipelines with explicit nodes and edges. You know, the kind that looks like this:
+
+```tsx
+const graph = new Graph()
+  .addNode("fetchHNPosts", fetchHNPosts)
+  .addNode("analyzeHNPosts", analyzePosts)
+  .addNode("generateReport", generateReport)
+  .addNode("editReport", editReport)
+  .addNode("writeTweet", writeTweet);
+
+graph
+  .addEdge(START, "fetchHNPosts")
+  .addEdge("fetchHNPosts", "analyzeHNPosts")
+  .addEdge("analyzeHNPosts", "generateReport")
+  .addEdge("generateReport", "editReport")
+  .addEdge("editReport", "writeTweet")
+  .addEdge("writeTweet", END);
+```
+
+The problem? These graph structures force you to juggle multiple mental models simultaneously. You've got the node definitions in one place, edge connections somewhere else, a separate execution flow to consider, and usually a global state object being passed around. And that's before we even get to complex patterns like branching, conditional execution, or—heaven forbid—loops.
+
+## Trees: The Elegant Simplification
+
+When we built GenSX, we made a radical bet: what if we used JSX to represent workflows as trees instead? Not because we're React fanboys (though some of us might be), but because tree structures match how humans naturally think about processes.
+
+Here's that same workflow in GenSX:
+
+```tsx
+<FetchHNPosts limit={postCount}>
+  {(stories) => (
+    <AnalyzeHNPosts stories={stories}>
+      {({ analyses }) => (
+        <GenerateReport analyses={analyses}>
+          {(report) => (
+            <EditReport content={report}>
+              {(editedReport) => (
+                <WriteTweet
+                  context={editedReport}
+                  prompt="Summarize the HN trends in a tweet"
+                />
+              )}
+            </EditReport>
+          )}
+        </GenerateReport>
+      )}
+    </AnalyzeHNPosts>
+  )}
+</FetchHNPosts>
+```
+
+"But wait," I can hear you protesting, "where's the graph?!"
+
+It's right there. A tree is just a particular kind of graph—one where each node has exactly one parent. And that constraint buys us immense clarity without sacrificing power.
+
+## When Trees Aren't Enough, Use JavaScript
+
+Trees work beautifully for most linear workflows. But what about loops, conditionals, and cycles? Those places where traditional graphs flex their muscles?
+
+Here's where GenSX's approach really shines. Instead of inventing some custom DSL for these patterns, we use... wait for it... regular JavaScript!
+
+Need to implement an agent that loops until a condition is met? Easy:
+
+```tsx
+const AgentWorkflow = gensx.Component<{}, AgentWorkflowOutput>(
+  "AgentWorkflow",
+  <AgentStep>
+    {(result) =>
+      result.needsMoreWork ? (
+        <AgentWorkflow /> // Recursion creates a cycle
```

Comment:
I wonder if we should show for loops rather than recursion here since the theme is "graphs are simpler". I know you mentioned that was one of your takeaways from the conference last week. We also mention recursion somewhere below in this post.

Reply:
Good call, I'll change this to a `while` loop.